FLEDGE.md

```diff
@@ -246,7 +246,7 @@ anywhere in the request where a plain `adRenderId` would have been sent (such as
 and `adComponents` fields as well as `prevWins`). Note that `include-full-ads` is not compatible
 with the auction server, so this mode is only for debugging.
 
-All fields that accept arbitrary metadata objects (`userBiddingSignals` and `metadata` field of ads) must be JSON-serializable.
+All fields that accept arbitrary metadata (`userBiddingSignals` and `metadata` field of ads) must be JSON-serializable values (which may include primitive values, arrays, objects of arbitrary structure).
```

Comment:
I'm not sure we want to say "objects of arbitrary structure" here as there are things in objects like functions that are not JSON-serializable.

Reply:
Yes. How about...

must be JSON-serializable (which may include arbitrary combinations or structures of primitive values, arrays, or objects).

Reply:
FYI, IIRC, another object that wouldn't be JSON serializable is an object with a circular reference:

```
let a = {};
let b = {'a': a};
a.b = b;
```

In the above example, neither `a` nor `b` are serializable.

Also, we might have some issues with really really deep nesting, like `[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[1]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]` (the actual max depth limit is 200 [0]). This only affects interest group updates, which use a non-V8-based recursive parser for JSON parsing of the update response.

Effectively, we're using the same operation as `JSON.stringify()` [1], so the rules documented in [2] should apply (note there are also rules around `BigInt`, `Date`, `Symbol`, etc.).

[0] https://source.chromium.org/chromium/chromium/src/+/main:base/json/json_common.h;l=18;drc=e4622aaeccea84652488d1822c28c78b7115684f

[1] https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify

[2] https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#description

Reply:
So, perhaps the explainer here should just link to the `JSON.stringify()` rules page on MDN.

Reply:
Thank you @caraitto.

So another try:

must be JSON-serializable as supported by `JSON.stringify()`. See [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify).

Reply:
I think "JSON-serializable values" and "supported by `JSON.stringify()`" are fairly redundant.  How about "JSON-serializable values (i.e. supported by [`JSON.stringify()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#description))"

Reply:
Sounds good. I'll update.

Reply:
Maybe link to https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify instead of https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#description?

I'm worried the note about circular references and BigInt might not get noticed since it's above the https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#description section.